src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,316 @@
+
+            // Log which sort option was used
+            Logger.Log(sortOption switch
+            {
+                SortOption.Name => FunctionId.DocumentOutline_SortByName,
+                SortOption.Order => FunctionId.DocumentOutline_SortByOrder,
+                SortOption.Type => FunctionId.DocumentOutline_SortByType,
+                _ => throw new NotImplementedException(),
+            });
+
+            return SortDocumentSymbolItems(documentSymbolItems, sortOption, cancellationToken);
+
+            static ImmutableArray<DocumentSymbolItem> SortDocumentSymbolItems(
+                ImmutableArray<DocumentSymbolItem> documentSymbolItems,
+                SortOption sortOption,
+                CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Sort the top-level DocumentSymbolItems
+                var sortedDocumentSymbolItems = sortOption switch
+                {
+                    SortOption.Name => documentSymbolItems.Sort((x, y) => x.Name.CompareTo(y.Name)),
+                    SortOption.Order => documentSymbolItems.Sort((x, y) =>
+                    {
+                        if (x.StartPosition.Line == y.StartPosition.Line)
+                            return x.StartPosition.Character - y.StartPosition.Character;
+
+                        return x.StartPosition.Line - y.StartPosition.Line;
+                    }),
+                    SortOption.Type => documentSymbolItems.Sort((x, y) =>
+                    {
+                        if (x.SymbolKind == y.SymbolKind)
+                            return x.Name.CompareTo(y.Name);
+
+                        return x.SymbolKind - y.SymbolKind;
+                    }),
+                    _ => throw new NotImplementedException()
+                };
+
+                // Recursively sort descendant DocumentSymbolItems
+                foreach (var documentSymbolItem in sortedDocumentSymbolItems)
+                {
+                    documentSymbolItem.Children = SortDocumentSymbolItems(documentSymbolItem.Children, sortOption, cancellationToken);
+                }
+
+                return sortedDocumentSymbolItems;
+            }
+        }
+
+        /// <summary>
+        /// Returns an immutable array of DocumentSymbolItem such that each node or one of its descendants matches the given pattern.
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolItem> Search(
+            ImmutableArray<DocumentSymbolItem> documentSymbolItems,
+            string pattern,
+            CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var documentSymbols = ArrayBuilder<DocumentSymbolItem>.GetInstance();
+            var patternMatcher = PatternMatcher.CreatePatternMatcher(pattern, includeMatchedSpans: false, allowFuzzyMatching: true);
```

Comment:
interesting, what do you use the matched spans for?  do we highlight hte matchign part of the name?